src/urllib3/response.py

```diff
@@ -223,6 +225,62 @@ def _get_decoder(mode: str) -> ContentDecoder:
     return DeflateDecoder()
 
 
+class BytesQueueBuffer:
+    """Memory-efficient bytes buffer
+
+    To return decoded data in read() and still follow the BufferedIOBase API, we need a
+    buffer to always return the correct amount of bytes.
+
+    This buffer should be filled using calls to put()
+
+    Our maximum memory usage is determined by the sum of the size of:
+
+     * self.buffer, which contains the full data
+     * the largest chunk that we will copy in get()
+
+    The worst case scenario is a single chunk, in which case we'll make a full copy of
+    the data inside get().
+    """
+
+    def __init__(self) -> None:
+        self.buffer: Deque[bytes] = collections.deque()
+        self._size: int = 0
+
+    def __len__(self) -> int:
+        return self._size
+
+    def put(self, data: bytes) -> None:
+        self.buffer.append(data)
+        self._size += len(data)
+
+    def get(self, n: int) -> bytes:
+        if not self.buffer:
+            raise ValueError("buffer is empty")
```

Comment:
nit: Should this be a `RuntimeError` instead of `ValueError`?

Reply:
Addressed in [`0dc26a3` (#2798)](https://github.com/urllib3/urllib3/pull/2798/commits/0dc26a352d3178f8d66784443f79ed65d18b60f6).